File: java/JumpGameII.java
public class JumpGameII {
    public int jump(int[] A) {
        if (A == null) {
            return -1;
        }
        int n = A.length;
        int steps = 0;
        int last = 0;
        int current = 0;
        for (int i = 0; i < n; ++i) {
            if (i > last) {
                if (last == current && current < A[n - 1] - 1) {
                    return -1;
                }
                last = current;
                ++steps;
            }
            current = Math.max(current, i + A[i]);
        }
        return steps;
    }
}

File: python/jump_game_ii.py
#!/usr/bin/env python3

class Solution:
    def jump(self, A):
        if A is None:
            return -1
        res = 0
        n = len(A)
        last = 0
        curr = 0
        for i in range(n):
            if i > last:
                last = curr
                res += 1
            curr = max(curr, i + A[i])
        return res

def main():
    solver = Solution()
    print(solver.jump([2, 3, 1, 1, 4]))


if __name__ == '__main__':
    main()
